File: csharp/CSharpToCypher/Cypher.cs
using System.Globalization;
using System.Text;

namespace CSharpToCypher;

/// <summary>Cypher emission: escaping, value formatting and the graph model.</summary>
public static class Cypher
{
    public static string EscapeString(string value)
    {
        var sb = new StringBuilder(value.Length);
        foreach (var ch in value)
        {
            switch (ch)
            {
                case '\\': sb.Append("\\\\"); break;
                case '\'': sb.Append("\\'"); break;
                case '"': sb.Append("\\\""); break;
                case '\r': sb.Append("\\r"); break;
                case '\n': sb.Append("\\n"); break;
                case '\t': sb.Append("\\t"); break;
                default: sb.Append(ch); break;
            }
        }
        return sb.ToString();
    }

    public static string FormatValue(object? value)
    {
        return value switch
        {
            null => "null",
            bool b => b ? "true" : "false",
            int i => i.ToString(CultureInfo.InvariantCulture),
            long l => l.ToString(CultureInfo.InvariantCulture),
            double d => d.ToString(CultureInfo.InvariantCulture),
            string s => "'" + EscapeString(s) + "'",
            IEnumerable<object?> list => "[" + string.Join(", ", list.Select(FormatValue)) + "]",
            _ => "'" + EscapeString(value.ToString() ?? "") + "'",
        };
    }
}

public sealed class CypherNode
{
    public required string Id { get; init; }
    public required IReadOnlyList<string> Labels { get; init; }
    public Dictionary<string, object?> Properties { get; } = new();

    public string ToCypher()
    {
        var props = new Dictionary<string, object?>(Properties) { ["id"] = Id };
        var propText = string.Join(", ", props
            .Where(kv => kv.Value is not null)
            .Select(kv => $"{kv.Key}: {Cypher.FormatValue(kv.Value)}"));
        return $"CREATE ({Id}:{string.Join(":", Labels)} {{{propText}}});";
    }
}

public sealed class CypherRelationship
{
    public required string Source { get; init; }
    public required string Type { get; init; }
    public required string Target { get; init; }
    public Dictionary<string, object?>? Properties { get; init; }

    public string ToCypher()
    {
        var props = Properties;
        var propText = props is { Count: > 0 }
            ? " {" + string.Join(", ", props.Select(kv => $"{kv.Key}: {Cypher.FormatValue(kv.Value)}")) + "}"
            : "";
        return $"CREATE ({Source})-[:{Type}{propText}]->({Target});";
    }
}

public sealed class CypherDocument
{
    public List<CypherNode> Nodes { get; } = new();
    public List<CypherRelationship> Relationships { get; } = new();
}

File: python/cypher_extractor/cypher.py
"""Cypher emission: escaping helpers plus the in-memory graph model.

The graph model is intentionally language-agnostic: a ``Node`` carries one or
more labels (the language-endemic construct names) and a property map; a
``Relationship`` carries the relationship type and optional properties. Both
know how to render themselves as Neo4j Cypher ``CREATE`` statements.
"""

from __future__ import annotations


def escape_string(value: str) -> str:
    """Escape a string for use inside a single-quoted Cypher string literal."""
    return (
        value.replace("\\", "\\\\")
        .replace("'", "\\'")
        .replace("\r", "\\r")
        .replace("\n", "\\n")
        .replace("\t", "\\t")
    )


def format_value(value):
    """Render a Python value as a Cypher literal."""
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return repr(value)
    if isinstance(value, str):
        return "'" + escape_string(value) + "'"
    if isinstance(value, (list, tuple)):
        return "[" + ", ".join(format_value(v) for v in value) + "]"
    return "'" + escape_string(str(value)) + "'"


class Node:
    """A Cypher node. ``labels`` may be a single string or an iterable of them."""

    __slots__ = ("id", "labels", "properties")

    def __init__(self, node_id, labels, properties=None):
        if isinstance(labels, str):
            labels = [labels]
        self.id = node_id
        self.labels = list(labels)
        self.properties = dict(properties or {})

    def to_cypher(self) -> str:
        props = dict(self.properties)
        props["id"] = self.id
        label_text = ":".join(self.labels)
        prop_text = ", ".join(f"{k}: {format_value(v)}" for k, v in props.items())
        return f"CREATE ({self.id}:{label_text} {{{prop_text}}});"


class Relationship:
    """A directed Cypher relationship between two node ids."""

    __slots__ = ("source", "type", "target", "properties")

    def __init__(self, source, type_, target, properties=None):
        self.source = source
        self.type = type_
        self.target = target
        self.properties = dict(properties or {})

    def to_cypher(self) -> str:
        if self.properties:
            prop_text = (
                " {"
                + ", ".join(f"{k}: {format_value(v)}" for k, v in self.properties.items())
                + "}"
            )
        else:
            prop_text = ""
        # Node and relationship statements are separate, independently
        # autocommitted CREATE statements (see Node.to_cypher), so a bare
        # variable like ``p1`` is not bound here - it would silently create a
        # fresh, disconnected node instead of referencing the one created
        # earlier. Look both endpoints up by their ``id`` property instead.
        source_id = format_value(self.source)
        target_id = format_value(self.target)
        return (
            f"MATCH (a {{id: {source_id}}}), (b {{id: {target_id}}}) "
            f"CREATE (a)-[:{self.type}{prop_text}]->(b);"
        )
